secrets/secrets-kubernetes/pkg/secrets/server.go

```diff
@@ -0,0 +1,218 @@
+package secrets
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nkeys"
+)
+
+type (
+	ServerErrorCallback  func(msg *nats.Msg, err error)
+	ServerContextCreator func() context.Context
+)
+
+type Server struct {
+	queue         *nats.Subscription
+	natsConn      *nats.Conn
+	handler       Handler
+	onError       ServerErrorCallback
+	key           nkeys.KeyPair
+	pubKey        string
+	subjectMapper SubjectMapper
+	ctxCreator    ServerContextCreator
+}
+
+type ServerOption func(*Server) error
+
+func WithEphemeralKey() ServerOption {
+	return func(s *Server) error {
+		var err error
+		s.key, err = nkeys.CreateCurveKeys()
+		return err
+	}
+}
+
+func WithKeyPair(kp nkeys.KeyPair) ServerOption {
+	return func(s *Server) error {
+		s.key = kp
+		return nil
+	}
+}
+
+func WithSubjectMapper(m SubjectMapper) ServerOption {
+	return func(s *Server) error {
+		s.subjectMapper = m
+		return nil
+	}
+}
+
+func WithErrorCallback(cb ServerErrorCallback) ServerOption {
+	return func(s *Server) error {
+		s.onError = cb
+		return nil
+	}
+}
+
+func WithRequestContext(cb ServerContextCreator) ServerOption {
+	return func(s *Server) error {
+		s.ctxCreator = cb
+		return nil
+	}
+}
+
+func NewServer(name string, nc *nats.Conn, handler Handler, opts ...ServerOption) (*Server, error) {
+	server := &Server{
+		natsConn:   nc,
+		handler:    handler,
+		onError:    func(*nats.Msg, error) {},
+		ctxCreator: func() context.Context { return context.Background() },
+		subjectMapper: SubjectMapper{
+			Version:     DefaultSecretsProtocolVersion,
+			Prefix:      DefaultSecretsBusPrefix,
+			ServiceName: name,
+		},
+	}
+
+	for _, opt := range opts {
+		if err := opt(server); err != nil {
+			return nil, fmt.Errorf("%w: %s", ErrInvalidServerConfig, err)
+		}
+	}
+
+	if name == "" {
+		return nil, fmt.Errorf("%w: missing name", ErrInvalidServerConfig)
+	}
+
+	if server.natsConn == nil {
+		return nil, fmt.Errorf("%w: nats connection", ErrInvalidServerConfig)
+	}
+
+	if server.handler == nil {
+		return nil, fmt.Errorf("%w: missing handler", ErrInvalidServerConfig)
+	}
+
+	if server.key == nil {
+		return nil, fmt.Errorf("%w: missing key pair", ErrInvalidServerConfig)
+	}
+
+	if server.ctxCreator == nil {
+		return nil, fmt.Errorf("%w: context creator", ErrInvalidServerConfig)
+	}
+
+	var err error
+	server.pubKey, err = server.key.PublicKey()
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrInvalidServerConfig, err)
+	}
+
+	return server, nil
+}
+
+func (s *Server) Process(ctx context.Context, msg *nats.Msg) {
+	nakCallback := func(reason error) {
+		s.onError(msg, reason)
+
+		if err := s.natsConn.Publish(msg.Reply, []byte(reason.Error())); err != nil {
```

Comment:
Btw, the nats.Msg type has a nice `respond` function for replying to a message easily https://pkg.go.dev/github.com/nats-io/nats.go#Msg.Respond. This might save you the `nakCallback` creation, but it may just end up being about the same /shrug

Reply:
this was a strange one. I was using `Respond` and getting a valid response with `-NAK` as Data.
made it look like i was interacting with a jetstream.
will check again 👍

Reply:
fixed!